### PR TITLE
Add quorum option to nats registry

### PR DESCRIPTION
**Pull request by phifty**

### CODE REVIEW

File: selector/blacklist/blacklist.go [hunk 1]
```diff
@@ -86,6 +86,26 @@ func (r *blacklistSelector) String() string {
 	return "blacklist"
 }
 
+func NewSelector(opts ...selector.Option) selector.Selector {
```
Comment (asim) [inline selector/blacklist/blacklist.go]: This should not be part of the same pull request.

File: registry/nats/nats.go [hunk 2]
```diff
@@ -31,6 +31,7 @@ var (
 	WatchTopic = "micro.registry.nats.watch"
 
 	DefaultTimeout = time.Millisecond * 100
+	DefaultQuorum  = 1
```
Comment (asim) [inline registry/nats/nats.go]: DefaultQuorum basically assumes if there is 1 result then it should return. This assumption is wrong. By default it should attempt to return all nodes within the given time limit.

File: registry/nats/nats.go [hunk 3]
```diff
@@ -338,6 +339,7 @@ func (n *natsRegistry) String() string {
 func NewRegistry(opts ...registry.Option) registry.Registry {
 	options := registry.Options{
 		Timeout: DefaultTimeout,
+		Quorum:  DefaultQuorum,
```
Comment (asim) [inline registry/nats/nats.go]: DefaultQuorum should be 0 and use timeout to return all nodes.

File: registry/nats/options.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
+package nats
+
+import (
+	"github.com/micro/go-micro/registry"
+	"golang.org/x/net/context"
+)
+
+const quorumKey = "quorum"
```
Comment (asim) [inline registry/nats/options.go]: quorumKey should be a struct like so https://github.com/micro/go-plugins/blob/master/registry/gossip/options.go#L9

File: registry/nats/options.go [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+var DefaultQuorum = 0
+
+func Quorum(n int) registry.Option {
+	return func(o *registry.Options) {
+		o.Context = context.WithValue(o.Context, quorumKey, n)
+	}
+}
+
+func GetQuorum(o *registry.Options) int {
```
Comment (asim) [inline registry/nats/options.go]: This doesn't need to be a public function and should also just take a non pointer.

File: registry/nats/nats.go [hunk 6]
```diff
@@ -236,13 +231,18 @@ loop:
...
+				serviceMap[key] = srv
+			} else {
 				serviceMap[key] = service
-				continue
 			}
-			srv.Nodes = append(srv.Nodes, service.Nodes...)
-			serviceMap[key] = srv
-		case <-done:
+
+			// log.Printf("nodes %d", len(serviceMap[key].Nodes))
```
Comment (asim) [inline registry/nats/nats.go]: unnecessary comment

File: registry/nats/options.go (hunk 4 above)
Comment (phifty) [inline registry/nats/options.go]: Could you give some background why a struct is used there?

Comment (asim) [inline registry/nats/options.go]: It's a pattern I picked up, I think from one of the Go team. Structs are unique to a package and provide clean and simple initialisation.

**Comment by asim** [comment]
Thanks for the contribution. The change looks good. On reflecting, I like this feature and think it adds great value to the nats registry. Merging now.

**Comment by phifty** [comment]
You're welcome. Thanks for the quick responses.

**Pull request merged**